portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -49,168 +49,203 @@
-        URL url = new URL(imagesService.getServingUrl(options));
-        return url.getPath();
-      } catch (MalformedURLException e) {
-        return imagesService.getServingUrl(options);
+ 
+  // HELPER FUNCTION:................................................................
+  // Loads all BlogMessages from Datastore if true is passed,
+  // Otherwise if false is passed only the recent post is loaded.
+  // This is useful because each time a user posts, we only load the last BlogMessage
+  // Which reduces the time to load datastore.
+  private List<BlogMessage> LoadAllBlogsOrLast(boolean all) {
+    List<BlogMessage> blogMessages = new ArrayList<BlogMessage>();
+    Query query = new Query(BLOG_ENTITY_KIND);
+ 
+    if (all) {
+      query.addSort("time", SortDirection.ASCENDING);
+    } else {
+      query.addSort("time", SortDirection.DESCENDING);
+    }
+ 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+ 
+    UserService userService = UserServiceFactory.getUserService();
+ 
+    for (Entity entity : results.asIterable()) {
+      long messageId = entity.getKey().getId();
+      long timestamp = (long) entity.getProperty("time");
+      String tag = (String) entity.getProperty("tag");
+      String comment = (String) entity.getProperty("text");
+      String nickname = (String) entity.getProperty("nickname");
+      String email = (String) userService.getCurrentUser().getEmail();
+      long parentID = (long) entity.getProperty("parentID");
+      ArrayList<BlogMessage> messageReplies = (ArrayList) entity.getProperty("replies");
+ 
+      BlogMessage message = new BlogMessage(
+            messageId, tag, comment, nickname, email, messageReplies, timestamp, parentID);
+      blogMessages.add(message);
+      if (!all) {
+        break;
       }
     }
-
-}
\ No newline at end of file
+    return blogMessages;
+  }
+ 
+  // HELPER FUNCTION:........................................................................
+  // Puts BlogMessages in BlogHashMap and loads the requested parameters.
+  private LinkedList<BlogMessage> sortAndLoadFromBlogHashMap(
+        List<BlogMessage> blogMessages, List<String> tagsToSearch, int loadAmount) {
+    BlogHashMap blogMap = new BlogHashMap();
+    blogMap.putInMap(blogMessages);
+ 
+    return blogMap.getMessages(tagsToSearch, loadAmount);
+  }
+ 
+  // HELPER FUNCTION:........................................................................
+  // Returns a URL that points to the uploaded file, or null if the user didn't upload a file.
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+ 
+    if (blobKeys == null || blobKeys.isEmpty()) {
```

Comment:
Move this check to the very beginning, even before you define those local variables.